datafusion/physical-expr/src/expressions/case.rs

```diff
@@ -122,6 +123,276 @@ fn is_cheap_and_infallible(expr: &Arc<dyn PhysicalExpr>) -> bool {
+            data_type: data_type.clone(),
+            partial_result_indices: vec![0; capacity],
+            partial_results: vec![],
+            covering_result: None,
+        }
+    }
+
+    /// Adds a result for one branch of the case expression.
+    ///
+    /// `row_indices` should be a [UInt32Array] containing [RecordBatch] relative row indices
+    /// for which `value` contains result values.
+    ///
+    /// If `value` is a scalar, the scalar value will be used as the value for each row in `row_indices`.
+    ///
+    /// If `value` is an array, the values from the array and the indices from `row_indices` will be
+    /// processed pairwise. The lengths of `value` and `row_indices` must match.
+    ///
+    /// The diagram below shows a situation where a when expression matched rows 1 and 4 of the
+    /// record batch. The then expression produced the value array `[A, D]`.
+    /// After adding this result, the result array will have been added to `partial_results` and
+    /// `partial_indices` will have been updated at indexes 1 and 4.
+    ///
+    /// ```text
+    /// ┌─────────┐     ┌─────────┐┌───────────┐                            ┌─────────┐┌───────────┐
+    /// │    A    │     │    0    ││           │                            │    0    ││┌─────────┐│
+    /// ├─────────┤     ├─────────┤│           │                            ├─────────┤││    A    ││
+    /// │    D    │     │    0    ││           │                            │    1    ││├─────────┤│
+    /// └─────────┘     ├─────────┤│           │   add_branch_result(       ├─────────┤││    D    ││
+    ///   value         │    0    ││           │     row indices,           │    0    ││└─────────┘│
+    ///                 ├─────────┤│           │     value                  ├─────────┤│           │
+    ///                 │    0    ││           │   )                        │    0    ││           │
+    /// ┌─────────┐     ├─────────┤│           │ ─────────────────────────▶ ├─────────┤│           │
+    /// │    1    │     │    0    ││           │                            │    1    ││           │
+    /// ├─────────┤     ├─────────┤│           │                            ├─────────┤│           │
+    /// │    4    │     │    0    ││           │                            │    0    ││           │
+    /// └─────────┘     └─────────┘└───────────┘                            └─────────┘└───────────┘
+    /// row indices
+    ///                   partial     partial                                 partial     partial
+    ///                   indices     results                                 indices     results
+    /// ```
+    fn add_branch_result(
+        &mut self,
+        row_indices: &ArrayRef,
+        value: ColumnarValue,
+    ) -> Result<()> {
+        match value {
+            ColumnarValue::Array(a) => {
+                assert_eq!(a.len(), row_indices.len());
+                if row_indices.len() == self.partial_result_indices.len() {
```

Comment:
I expected this also to have to check the values in row_indices to make sure all rows were covered. Does this assume that there are no duplicates in row_indices?

Reply:
Yes, the code very much makes that assumption. It's tightly coupled to the way `case_when_with_expr` and `case_when_no_expr` are implemented. We could add validation checks here, but you end up spending quite some time verifying invariants. I couldn't figure out a way to check this without introducing overhead.